### PR TITLE
refactor: refactor table metadata manager

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

refactor table metadata manager

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#1715

### CODE REVIEW

File: src/common/meta/src/key.rs [hunk 1]
```diff
@@ -167,6 +189,275 @@ impl TableMetadataManager {
...
+
+        // Creates datanode table key value pairs.
+        let distribution = region_distribution(&region_routes)?;
+        self.datanode_table_manager()
+            .build_create_txn(&mut txn, table_id, distribution)?;
+
+        // Creates table route.
+        let table_route_value = TableRouteValue::new(region_routes);
+        self.table_route_manager()
+            .build_create_txn(&mut txn, table_id, &table_route_value)?;
```
Comment (fengjiachun) [inline src/common/meta/src/key.rs:229]: I prefer to build Txn in this way:
```
let (xxx_compares, xxx_successes) = build_xxx();
let (xx_compares, xx_successes) = build_xx();

// merge all 

let txn = Txn::new().when(all_compares).and_then(all_successes);
```

What do you think?

File: src/common/meta/src/key.rs [hunk 2]
```diff
@@ -167,6 +189,275 @@ impl TableMetadataManager {
...
+
+        let r = self.kv_backend.txn(txn.into()).await?;
+
+        // Checks whether metadata was already created.
+        if !r.succeeded {
```
Comment (fengjiachun) [inline src/common/meta/src/key.rs:234]: Why not use txn's or_else?

File: src/common/meta/src/key/datanode_table.rs [hunk 3]
```diff
@@ -105,6 +106,96 @@ impl DatanodeTableManager {
...
+        &self,
+        txn: &mut TxnRequest,
+        table_id: TableId,
+        current_region_distribution: RegionDistribution,
+        new_region_distribution: RegionDistribution,
+    ) -> Result<()> {
+        let mut opts = Vec::new();
+
+        // Removes the old datanode table key value pairs
+        for current_datanode in current_region_distribution.keys() {
```
Comment (MichaelScofield) [inline src/common/meta/src/key/datanode_table.rs:142]: The txn seems fairly complicated, please add more tests.

File: src/common/meta/src/key/datanode_table.rs [hunk 4]
```diff
@@ -105,6 +106,96 @@ impl DatanodeTableManager {
...
+    ) -> Result<()> {
+        let txns = distribution
+            .into_iter()
+            .map(|(datanode_id, regions)| {
+                let key = DatanodeTableKey::new(datanode_id, table_id);
+                let raw_key = key.as_raw_key();
+                let removed_key = to_removed_key(&String::from_utf8_lossy(&raw_key));
+                let val = DatanodeTableValue::new(table_id, regions);
+
+                Ok(TxnOp::Put(
```
Comment (MichaelScofield) [inline src/common/meta/src/key/datanode_table.rs:187]: need to delete the original key

File: src/common/meta/src/key.rs [hunk 5]
```diff
@@ -167,6 +189,275 @@ impl TableMetadataManager {
     pub fn schema_manager(&self) -> &SchemaManager {
         &self.schema_manager
     }
+
+    pub fn table_route_manager(&self) -> &TableRouteManager {
+        &self.table_route_manager
+    }
+
+    /// Creates metadata for table and returns an error if different metadata exists.
+    /// The caller MUST ensure it has the exclusive access to `TableNameKey`.
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key.rs:198]: The procedure's lock

File: src/common/meta/src/key.rs [hunk 6]
```diff
@@ -167,6 +189,275 @@ impl TableMetadataManager {
...
+        // Creates table route.
+        let table_route_value = TableRouteValue::new(region_routes);
+        self.table_route_manager()
+            .build_create_txn(&mut txn, table_id, &table_route_value)?;
+
+        let r = self.kv_backend.txn(txn.into()).await?;
+
+        // Checks whether metadata was already created.
+        if !r.succeeded {
+            let mut batch = BatchGetRequest::default();
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key.rs:235]: Yes, I'm refactoring it.

File: src/common/meta/src/key/datanode_table.rs [hunk 7]
```diff
@@ -105,6 +106,96 @@ impl DatanodeTableManager {
...
+            .into_iter()
+            .map(|(datanode_id, regions)| {
+                let key = DatanodeTableKey::new(datanode_id, table_id);
+                let val = DatanodeTableValue::new(table_id, regions);
+
+                Ok(TxnOp::Put(key.as_raw_key(), val.try_as_raw_value()?))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        txn.success.extend(txns);
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key/datanode_table.rs:126]: No need, `DatanodeTable` is just the cache of the `TableRoute`

File: src/common/meta/src/key/datanode_table.rs (hunk 4 above)
Comment (WenyXu) [reply src/common/meta/src/key/datanode_table.rs:187]: Same reason, `DatanodeTable` is just the cache of the `TableRoute`. We can recover all `DatanodeTable` from the removed `TableRoute` key.

**Comment by WenyXu** [comment]
@MichaelScofield @fengjiachun PTAL

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/common/meta/src/kv_backend/txn.rs [hunk 8]
```diff
@@ -121,6 +129,21 @@ pub struct Txn {
 }
 
 impl Txn {
+    pub fn merge_all<T: IntoIterator<Item = Txn>>(values: T) -> Self {
+        // Safety: values must contain an element.
+        values.into_iter().reduce(|acc, e| acc.merge(e)).unwrap()
```
Comment (fengjiachun) [inline src/common/meta/src/kv_backend/txn.rs:134]: Can we use `unwrap_or_default()` ? It seems that it is not possible to restrict the iterator to contain at least one element.

**Review by fengjiachun — APPROVED** [review]

**Comment by WenyXu** [comment]
@MichaelScofield PTAL

File: src/common/meta/src/key.rs [hunk 9]
```diff
@@ -221,12 +511,282 @@ impl_table_meta_value! {
...
+        // if remote metadata was updated, it should be ok.
+        table_metadata_manager
+            .rename_table(table_info_value.clone(), new_table_name.clone())
+            .await
+            .unwrap();
+        let mut modified_table_info = table_info;
+        modified_table_info.name = "hi".to_string();
+        let modified_table_info_value = table_info_value.update(modified_table_info);
+        // if the table_info_value is wrong, it should return an error.
+        // The ABA problem.
```
Comment (MichaelScofield) [inline src/common/meta/src/key.rs:664]: Why it's an "ABA" problem? The key seems didn't change to another value and back, during the txn.

File: src/common/meta/src/key.rs [hunk 10]
```diff
@@ -221,12 +511,282 @@ impl_table_meta_value! {
...
+            .create_table_metadata(table_info.clone(), region_routes.clone())
+            .await
+            .unwrap();
+        let mut modified_region_routes = region_routes.clone();
+        modified_region_routes.push(region_route);
+        // if remote metadata was exists, it should return an error.
+        assert!(table_metadata_manager
+            .create_table_metadata(table_info, modified_region_routes)
+            .await
+            .is_err());
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key.rs:609]: I added in the next PRs

File: src/common/meta/src/key.rs (hunk 9 above)
Comment (WenyXu) [reply src/common/meta/src/key.rs:664]: Yes, but the comparing value changed.

File: src/common/meta/src/key.rs [hunk 11]
```diff
@@ -192,6 +462,13 @@ macro_rules! impl_table_meta_value {
     ($($val_ty: ty), *) => {
         $(
             impl $val_ty {
+                pub fn try_from_raw_value_ref(raw_value: &[u8]) -> Result<Self> {
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key.rs:465]: I'll refactor these methods in the following PRs.

File: src/common/meta/src/key.rs [hunk 12]
```diff
@@ -167,6 +188,255 @@ impl TableMetadataManager {
...
+            create_table_route_txn,
+        ]);
+
+        let r = self.kv_backend.txn(txn).await?;
+
+        // Checks whether metadata was already created.
+        if !r.succeeded {
+            let remote_table_info =
+                on_create_table_info_failure(&r.responses)?.context(error::UnexpectedSnafu {
+                    err_msg: "Reads the empty table info during the create table metadata",
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key.rs:246]: We only allow to create `TableRoute` and `TableInfo` in the same transaction. Therefore, it only has two cases. Both are empty, or both exist.

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**